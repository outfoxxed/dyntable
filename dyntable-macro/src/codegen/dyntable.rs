use proc_macro2::TokenStream;
use quote::format_ident;
use syn::{Signature, Token};

use crate::parse::{DynTraitInfo, VTableEntry};

pub fn impl_dyntable(dyntrait: &DynTraitInfo) -> TokenStream {
	let proxy_trait = format_ident!("__DynTable_{}", dyntrait.dyntrait.ident);
	let ident = &dyntrait.dyntrait.ident;
	let vtable_ident = &dyntrait.vtable.name;
	let (_, ty_generics, where_clause) = dyntrait.generics.split_for_impl();
	let impl_generics = dyntrait.generics.params.clone().into_iter()
    .collect::<Vec<_>>();
	let where_entries = where_clause
    .into_iter()
    .flat_map(|clause| &clause.predicates);

	let vtable_entries = dyntrait.entries.iter()
    .map(|entry| match entry {
			VTableEntry::Subtable(subtable) => {
				let entry_name = &subtable.ident;
				quote::quote! {
					#entry_name: <__DynTarget as ::dyntable::DynTable<
						<(dyn #ident #ty_generics + 'static) as ::dyntable::VTableRepr>::VTable,
					>>::VTABLE
				}
			},
			VTableEntry::Method(method) => {
				let Signature {
					unsafety,
					abi,
					fn_token,
					ident,
					inputs,
					variadic,
					output,
					..
				} = &method.sig;

				let inputs = inputs.iter()
					.map(|_| <Token![_]>::default());

				let output = match output {
					syn::ReturnType::Default => None,
					syn::ReturnType::Type(..) => Some(<Token![_]>::default()),
				}.into_iter();

				quote::quote! {
					#ident: unsafe {
						::core::intrinsics::transmute(
							__DynTarget::#ident as
								#unsafety #abi #fn_token (
									#(#inputs,)* #variadic
								) #( -> #output)*
						)
					}
				}
			},
		});

	quote::quote! {
		#[allow(non_camel_case_types)]
		unsafe trait #proxy_trait<'v, V: 'v + ::dyntable::VTable> {
			const VTABLE: V;
			const STATIC_VTABLE: &'v V;
		}

		unsafe impl<
			'__dyn_vtable,
			#(#impl_generics + '__dyn_vtable,)*
			__DynTarget,
		> ::dyntable::__private::DynTable2<'__dyn_vtable, #vtable_ident #ty_generics>
		for ::dyntable::__private::DynImplTarget<__DynTarget, #vtable_ident #ty_generics>
		where
			#(#where_entries,)*
			__DynTarget: #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>,
		{
			const VTABLE: #vtable_ident #ty_generics = __DynTarget::VTABLE;
			const STATIC_VTABLE: &'__dyn_vtable #vtable_ident #ty_generics = __DynTarget::STATIC_VTABLE;
		}

		unsafe impl<
			'__dyn_vtable,
			#(#impl_generics + '__dyn_vtable,)*
			__DynTarget: #ident #ty_generics,
		> #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>
		for __DynTarget
		#where_clause {
			const STATIC_VTABLE: &'__dyn_vtable #vtable_ident #ty_generics =
				&<Self as #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>>::VTABLE;
			const VTABLE: #vtable_ident #ty_generics = #vtable_ident {
				#(#vtable_entries,)*
			};
		}
	}
}
