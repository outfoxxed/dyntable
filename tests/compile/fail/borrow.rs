use dyntable::{dyntable, DynRefMut};

fn main() {}

#[dyntable]
trait TestTrait {}

fn _double_mut_borrow(mut dynref: DynRefMut<dyn TestTrait>) {
		let a = dynref.borrow_mut();
		let b = dynref.borrow_mut();
		let _ = (a, b);
}

fn _shared_mut_overlap1(mut dynref: DynRefMut<dyn TestTrait>) {
	let a = dynref.borrow();
	let b = dynref.borrow_mut();
	let _ = (a, b);
}

fn _shared_mut_overlap2(mut dynref: DynRefMut<dyn TestTrait>) {
	let a = dynref.borrow_mut();
	let b = dynref.borrow();
	let _ = (a, b);
}
