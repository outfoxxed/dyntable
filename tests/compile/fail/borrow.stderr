error[E0499]: cannot borrow `dynref` as mutable more than once at a time
  --> tests/compile/fail/borrow.rs:10:11
   |
9  |         let a = dynref.borrow_mut();
   |                 ------------------- first mutable borrow occurs here
10 |         let b = dynref.borrow_mut();
   |                 ^^^^^^^^^^^^^^^^^^^ second mutable borrow occurs here
11 |         let _ = (a, b);
   |                  - first borrow later used here

error[E0502]: cannot borrow `dynref` as mutable because it is also borrowed as immutable
  --> tests/compile/fail/borrow.rs:16:10
   |
15 |     let a = dynref.borrow();
   |             --------------- immutable borrow occurs here
16 |     let b = dynref.borrow_mut();
   |             ^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
17 |     let _ = (a, b);
   |              - immutable borrow later used here

error[E0502]: cannot borrow `dynref` as immutable because it is also borrowed as mutable
  --> tests/compile/fail/borrow.rs:22:10
   |
21 |     let a = dynref.borrow_mut();
   |             ------------------- mutable borrow occurs here
22 |     let b = dynref.borrow();
   |             ^^^^^^^^^^^^^^^ immutable borrow occurs here
23 |     let _ = (a, b);
   |              - mutable borrow later used here
